haproxy-healthcheck/check.py

```diff
@@ -1,80 +1,87 @@
+
+
+def _send_up_response(writer: asyncio.StreamWriter) -> None:
+    _send_and_close(writer, b"up\n")
 
-def threaded(conn):
-    try:
-        result = []
-        couchdb_server = couchdb_servers[0]
-        try:
-            r = requests.get(get_membership_endpoint(couchdb_server), timeout=1)
-        except Exception as e:
-            print("Exception: ", e)
-        data = r.json()
 
-        all_nodes = sorted(data['all_nodes'])
-        cluster_nodes = sorted(data['cluster_nodes'])
+def _send_and_close(writer: asyncio.StreamWriter, message: bytes) -> None:
+    print(f"Response: {message!r}")
+    writer.write(message)
+    print("Closing connection")
+    writer.close()
+
+
+async def is_healthy() -> bool:
+    """
+    Checks that all couchdb servers are part of the cluster
+    """
+    try:
+        first_couchdb_server: str = couchdb_servers[0]
+        async with httpx.AsyncClient() as client:
+            r = await client.get(
+                get_membership_endpoint(first_couchdb_server), timeout=1
+            )
+        data: dict = r.json()
+        all_nodes: List[str] = sorted(data["all_nodes"])
+        cluster_nodes: List[str] = sorted(data["cluster_nodes"])
 
         if len(all_nodes) != len(couchdb_servers):
-            print('Nodes starting up')
-            print(f'Details: all_nodes: {all_nodes}. couchdb_servers: {couchdb_servers}')
-            result.append(b'down\n')
+            print("Nodes starting up")
+            print(
+                f"Details: all_nodes: {all_nodes}. couchdb_servers: {couchdb_servers}"
+            )
+            return False
         elif all_nodes != cluster_nodes:
-            print('_membership shows not all nodes are part of Cluster')
-            print(f'Details: all_nodes: {all_nodes}. cluster_nodes: {cluster_nodes}')
-            result.append(b'down\n')
+            print("_membership shows not all nodes are part of Cluster")
+            print(f"Details: all_nodes: {all_nodes}. cluster_nodes: {cluster_nodes}")
+            return False
         else:
-            print('Everything is fine')
-            result.append(b'up\n')
-        print(result)
+            return True
+
     except Exception as e:
-        result.append(b'down\n')
-        print("Exception: ", e)
-    finally:
-        if b'down\n' in result:
-            send_down_response(conn)
-        else:
-            send_up_response(conn)
+        print("Exception when checking health:", e)
+        return False
+
+
+async def handle_healthcheck(
+    _: asyncio.StreamReader, writer: asyncio.StreamWriter
+) -> None:
+    if await is_healthy():
+        _send_up_response(writer)
+    else:
+        _send_down_response(writer)
+
+
+async def main() -> None:
+    server: asyncio.Server = await asyncio.start_server(
+        handle_healthcheck, "0.0.0.0", 5555
+    )
 
-    print_lock.release()
-    print('Client disconnected')
+    addrs: str = ", ".join(str(sock.getsockname()) for sock in server.sockets)
```

Comment:
Not sure how this is supposed to work and how frequently I'm supposed to see these requests. 
At this time, I'm seeing requests in clusters of three, every second: 
```
couchdb-1.local-1  | [notice] 2023-11-30T13:48:11.215642Z couchdb@couchdb-1.local <0.13750.0> cc6f80bc0f couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:11.346468Z couchdb@couchdb-1.local <0.13751.0> 113e7ff3ff couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:12.237689Z couchdb@couchdb-1.local <0.13762.0> 187b06a1e5 couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:16.412544Z couchdb@couchdb-1.local <0.13817.0> 629c8c907a couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:16.593743Z couchdb@couchdb-1.local <0.13818.0> 1d053bbfbf couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:17.385188Z couchdb@couchdb-1.local <0.13832.0> f3badc652d couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:21.428283Z couchdb@couchdb-1.local <0.13875.0> 6500a4056a couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:21.860183Z couchdb@couchdb-1.local <0.13883.0> b22a502ceb couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:22.595296Z couchdb@couchdb-1.local <0.13886.0> 95bbf4bd32 couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:26.252207Z couchdb@couchdb-1.local <0.13939.0> 4d632bf9f9 couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:26.988184Z couchdb@couchdb-1.local <0.13941.0> 9f90675e8d couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:27.668225Z couchdb@couchdb-1.local <0.13959.0> 136021c4d7 couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:31.214254Z couchdb@couchdb-1.local <0.13984.0> 903afd4397 couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:31.802219Z couchdb@couchdb-1.local <0.14005.0> 66873b7097 couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
couchdb-1.local-1  | [notice] 2023-11-30T13:48:32.554425Z couchdb@couchdb-1.local <0.14008.0> d257302f70 couchdb-1.local:5984 172.18.0.2 medic GET /_membership 200 ok 0
````
Are we supposed to issue more 3 requests per second like this?

Reply:
I would expect one request every 5 seconds which is what I'm seeing on my machine, I updated the PR with fixed logging in healthcheck, could you see if there's any outliers in that log? How often does the old healthcheck request for you? Do the healthchecks stop if you stop haproxy?

Reply:
I pulled the latest updates and this is what I'm seeing. ~3 requests per second. 

```
[notice] 2023-12-04T17:21:52.675936Z couchdb@couchdb-1.local <0.4400.0> d006485f4e couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:21:53.612867Z couchdb@couchdb-1.local <0.4406.0> 9ee8ea71c6 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:21:55.552293Z couchdb@couchdb-1.local <0.4431.0> a57c0543ea couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:21:57.720849Z couchdb@couchdb-1.local <0.4465.0> 60cf04e39c couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:21:58.775760Z couchdb@couchdb-1.local <0.4475.0> 28716f8691 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:00.374407Z couchdb@couchdb-1.local <0.4482.0> 22f7a55e26 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:02.921226Z couchdb@couchdb-1.local <0.4508.0> 36b4069d5e couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:03.801875Z couchdb@couchdb-1.local <0.4514.0> 375d738c99 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:05.217400Z couchdb@couchdb-1.local <0.4536.0> c9971cf9cd couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:08.192594Z couchdb@couchdb-1.local <0.4573.0> d2ce3f0809 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:08.843281Z couchdb@couchdb-1.local <0.4582.0> 24ef183a19 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:10.396254Z couchdb@couchdb-1.local <0.4590.0> 7d92748a0f couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:13.286201Z couchdb@couchdb-1.local <0.4617.0> 17e82a4632 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:13.846717Z couchdb@couchdb-1.local <0.4621.0> d2c4047586 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:15.424240Z couchdb@couchdb-1.local <0.4644.0> 62a11c5d65 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:18.106622Z couchdb@couchdb-1.local <0.4680.0> a1a2cb802b couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:18.805958Z couchdb@couchdb-1.local <0.4688.0> 51924bcad6 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:20.504615Z couchdb@couchdb-1.local <0.4702.0> 2c9e62947d couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:23.102532Z couchdb@couchdb-1.local <0.4722.0> 5dda6e2e43 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:25.430224Z couchdb@couchdb-1.local <0.4813.0> a5dbb84825 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:27.919870Z couchdb@couchdb-1.local <0.5136.0> 5e638db6d3 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:28.774652Z couchdb@couchdb-1.local <0.5142.0> 04448acb99 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:30.302203Z couchdb@couchdb-1.local <0.5153.0> 8dffac7331 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:32.980015Z couchdb@couchdb-1.local <0.5312.0> 4445f24d95 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:33.754985Z couchdb@couchdb-1.local <0.5317.0> bc467fea06 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:35.193795Z couchdb@couchdb-1.local <0.5347.0> 1dc1662b95 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:38.117570Z couchdb@couchdb-1.local <0.5386.0> f25899805d couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:38.706203Z couchdb@couchdb-1.local <0.5391.0> 0c62b39ca6 couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
[notice] 2023-12-04T17:22:40.259846Z couchdb@couchdb-1.local <0.5403.0> c43dd5be0b couchdb-1.local:5984 172.18.0.3 medic GET /_membership 200 ok 0
```

Reply:
Back at this with more logs: 

CouchDb: 
```
[notice] 2023-12-18T17:36:38.624186Z couchdb@couchdb-1.local <0.1484.0> 879bd90390 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:39.855767Z couchdb@couchdb-1.local <0.1565.0> 498d128df2 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:41.334425Z couchdb@couchdb-1.local <0.1637.0> 31e2e45f65 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:43.841258Z couchdb@couchdb-1.local <0.1658.0> ed7b303e20 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:44.764484Z couchdb@couchdb-1.local <0.1672.0> ff7544af07 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:46.353449Z couchdb@couchdb-1.local <0.1676.0> 188d80cb3e couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:48.659975Z couchdb@couchdb-1.local <0.1697.0> e8f9dc7ac0 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:49.722659Z couchdb@couchdb-1.local <0.1710.0> dd0330667a couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:51.543659Z couchdb@couchdb-1.local <0.1742.0> 114ace9018 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:53.460047Z couchdb@couchdb-1.local <0.1762.0> 0765124c62 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:54.726118Z couchdb@couchdb-1.local <0.1776.0> 200c8f3e78 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:56.265636Z couchdb@couchdb-1.local <0.425.0> -------- rexi_server : cluster stable
[notice] 2023-12-18T17:36:56.265684Z couchdb@couchdb-1.local <0.429.0> -------- rexi_buffer : cluster stable
[notice] 2023-12-18T17:36:56.652859Z couchdb@couchdb-1.local <0.1785.0> 306547b903 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:58.308536Z couchdb@couchdb-1.local <0.1796.0> b47fbb8953 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:36:59.590762Z couchdb@couchdb-1.local <0.1819.0> 3810353a1b couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:01.505556Z couchdb@couchdb-1.local <0.1847.0> d2f80def23 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:03.430498Z couchdb@couchdb-1.local <0.1866.0> dfdb4a97b9 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:04.843059Z couchdb@couchdb-1.local <0.1881.0> 8f6bd154da couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:06.366734Z couchdb@couchdb-1.local <0.1885.0> c1ea86fcf0 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:08.249733Z couchdb@couchdb-1.local <0.1900.0> 786ca3c3d0 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:09.895076Z couchdb@couchdb-1.local <0.1924.0> 398d274f04 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:11.265916Z couchdb@couchdb-1.local <0.1950.0> 92a6a1a524 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:13.415727Z couchdb@couchdb-1.local <0.1971.0> 6e466cc487 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:14.752613Z couchdb@couchdb-1.local <0.1985.0> 473b5f6248 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:16.276227Z couchdb@couchdb-1.local <0.1989.0> 253d271b66 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:18.521266Z couchdb@couchdb-1.local <0.2006.0> db1203f65f couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
[notice] 2023-12-18T17:37:19.566704Z couchdb@couchdb-1.local <0.2024.0> 58f8e53738 couchdb-1.local:5984 172.18.0.5 medic GET /_membership 200 ok 0
```

Healthcheck: 
```
2023-12-18 17:37:24,777 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:24,777 - root - INFO - Response: b'up\n'
2023-12-18 17:37:24,777 - root - INFO - Closing connection
2023-12-18 17:37:26,395 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:26,395 - root - INFO - Response: b'up\n'
2023-12-18 17:37:26,395 - root - INFO - Closing connection
2023-12-18 17:37:28,422 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:28,423 - root - INFO - Response: b'up\n'
2023-12-18 17:37:28,423 - root - INFO - Closing connection
2023-12-18 17:37:29,596 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:29,597 - root - INFO - Response: b'up\n'
2023-12-18 17:37:29,597 - root - INFO - Closing connection
2023-12-18 17:37:31,417 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:31,418 - root - INFO - Response: b'up\n'
2023-12-18 17:37:31,418 - root - INFO - Closing connection
2023-12-18 17:37:33,251 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:33,252 - root - INFO - Response: b'up\n'
2023-12-18 17:37:33,252 - root - INFO - Closing connection
2023-12-18 17:37:34,842 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:34,842 - root - INFO - Response: b'up\n'
2023-12-18 17:37:34,842 - root - INFO - Closing connection
2023-12-18 17:37:36,433 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:36,434 - root - INFO - Response: b'up\n'
2023-12-18 17:37:36,434 - root - INFO - Closing connection
2023-12-18 17:37:38,289 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:38,289 - root - INFO - Response: b'up\n'
2023-12-18 17:37:38,289 - root - INFO - Closing connection
2023-12-18 17:37:39,755 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:39,755 - root - INFO - Response: b'up\n'
2023-12-18 17:37:39,755 - root - INFO - Closing connection
2023-12-18 17:37:41,268 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:41,269 - root - INFO - Response: b'up\n'
2023-12-18 17:37:41,269 - root - INFO - Closing connection
2023-12-18 17:37:43,509 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:43,509 - root - INFO - Response: b'up\n'
2023-12-18 17:37:43,509 - root - INFO - Closing connection
2023-12-18 17:37:44,545 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:44,545 - root - INFO - Response: b'up\n'
2023-12-18 17:37:44,545 - root - INFO - Closing connection
2023-12-18 17:37:46,139 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:46,139 - root - INFO - Response: b'up\n'
2023-12-18 17:37:46,139 - root - INFO - Closing connection
2023-12-18 17:37:48,336 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:48,337 - root - INFO - Response: b'up\n'
2023-12-18 17:37:48,337 - root - INFO - Closing connection
2023-12-18 17:37:49,806 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:49,807 - root - INFO - Response: b'up\n'
2023-12-18 17:37:49,807 - root - INFO - Closing connection
2023-12-18 17:37:51,142 - httpx - INFO - HTTP Request: GET http://medic:password@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2023-12-18 17:37:51,143 - root - INFO - Response: b'up\n'
2023-12-18 17:37:51,143 - root - INFO - Closing connection
```

Reply:
@nydr 
I checked out your branch and failed one test deliberately, just so we have the artefact with the logs from the CI. You can find the logs here: https://github.com/medic/cht-core/actions/runs/7449664608

I'm still seeing requests more frequently than every 5 seconds: 
```
2024-01-08 15:50:41,304 - httpx - INFO - HTTP Request: GET http://admin:pass@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2024-01-08 15:50:41,305 - root - INFO - Response: b'up\n'
2024-01-08 15:50:41,305 - root - INFO - Closing connection
2024-01-08 15:50:43,040 - httpx - INFO - HTTP Request: GET http://admin:pass@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2024-01-08 15:50:43,040 - root - INFO - Response: b'up\n'
2024-01-08 15:50:43,041 - root - INFO - Closing connection
2024-01-08 15:50:44,285 - httpx - INFO - HTTP Request: GET http://admin:pass@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2024-01-08 15:50:44,286 - root - INFO - Response: b'up\n'
2024-01-08 15:50:44,287 - root - INFO - Closing connection
2024-01-08 15:50:46,281 - httpx - INFO - HTTP Request: GET http://admin:pass@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2024-01-08 15:50:46,282 - root - INFO - Response: b'up\n'
2024-01-08 15:50:46,282 - root - INFO - Closing connection
2024-01-08 15:50:48,289 - httpx - INFO - HTTP Request: GET http://admin:pass@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2024-01-08 15:50:48,289 - root - INFO - Response: b'up\n'
2024-01-08 15:50:48,290 - root - INFO - Closing connection
2024-01-08 15:50:49,212 - httpx - INFO - HTTP Request: GET http://admin:pass@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2024-01-08 15:50:49,213 - root - INFO - Response: b'up\n'
2024-01-08 15:50:49,213 - root - INFO - Closing connection
2024-01-08 15:50:51,233 - httpx - INFO - HTTP Request: GET http://admin:pass@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2024-01-08 15:50:51,234 - root - INFO - Response: b'up\n'
2024-01-08 15:50:51,234 - root - INFO - Closing connection
2024-01-08 15:50:53,474 - httpx - INFO - HTTP Request: GET http://admin:pass@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2024-01-08 15:50:53,475 - root - INFO - Response: b'up\n'
2024-01-08 15:50:53,475 - root - INFO - Closing connection
2024-01-08 15:50:54,467 - httpx - INFO - HTTP Request: GET http://admin:pass@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2024-01-08 15:50:54,468 - root - INFO - Response: b'up\n'
2024-01-08 15:50:54,468 - root - INFO - Closing connection
2024-01-08 15:50:56,214 - httpx - INFO - HTTP Request: GET http://admin:pass@couchdb-1.local:5984/_membership "HTTP/1.1 200 OK"
2024-01-08 15:50:56,215 - root - INFO - Response: b'up\n'
2024-01-08 15:50:56,215 - root - INFO - Closing connection
```

I wonder if this could be because every server is checked separately? 
When you are testing locally, how many couchdb containers do you launch?

Reply:
> @nydr I checked out your branch and failed one test deliberately, just so we have the artefact with the logs from the CI. You can find the logs here: https://github.com/medic/cht-core/actions/runs/7449664608

This is awesome and will help lots with reproducing, thanks! I'll aim to look at this next week.
> When you are testing locally, how many couchdb containers do you launch?

One, if I run three I'm seeing three requests per 5 seconds as expected.

_Note that all healthchecks only go to the first server just like the previous healthcheck did:_ https://github.com/medic/cht-core/blob/25bbec5ac73dc2d39d4df95d03a02e5b3cf28982/haproxy-healthcheck/check.py#L34

Reply:
Oh, I see. Healthcheck isn't needed at all in single node though. So I suppose this is ok, seeing 3 requests per 5 seconds.